drivers/rtc/rtc_nxp_irtc.c

```diff
@@ -0,0 +1,502 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/devicetree.h>
+#define DT_DRV_COMPAT nxp_irtc
+
+#include <zephyr/drivers/rtc.h>
+#include <zephyr/irq.h>
+
+struct nxp_irtc_config {
+	RTC_Type *base;
+	uint8_t alarm_match_flag;
+	uint8_t clock_select;
+	bool is_output_clock_enabled;
+	void (*irq_config_func)(const struct device *dev);
+};
+
+struct nxp_irtc_data {
+	bool is_dst_enabled;
+#ifdef CONFIG_RTC_ALARM
+	rtc_alarm_callback alarm_callback;
+	void *alarm_user_data;
+	uint16_t *alarm_mask;
+	bool is_alarm_pending;
+#endif /* CONFIG_RTC_ALARM */
+#ifdef CONFIG_RTC_UPDATE
+	rtc_update_callback update_callback;
+	void *update_user_data;
+#endif /* CONFIG_RTC_ALARM */
+#ifdef CONFIG_RTC_CALIBRATION
+	int32_t calibration;
+#endif /* CONFIG_RTC_CALIBRATION */
+};
+
+/* The IRTC Offset is 2112 instead of 1990 [2112 - 1990] -> [212] */
+#define RTC_NXP_IRTC_YEAR_OFFSET 212
+
+/* Utility function for testing if bit for alarm mask is enabled. */
+#define RTC_NXP_RTCTIME_COMPARE_FLAG(mask, flag) (((mask) == 0) || (((mask) & (flag)) && (flag)))
+
+/* Utility function to check the bounds of passed in rtc_time */
+static uint8_t is_rtc_time_in_bounds(const struct rtc_time *time_value, uint16_t mask)
```

Comment:
use the rtc_utils_validate_rtc_time in rtc_utils.h ?

Reply:
The bounds are different with the IRTC.

Reply:
which ones are different?

Reply:
The Year, I would hate to use two functions when only needing to cater to the special case of the year member in the struct.

Reply:
I agree with @decsny here, use `rtc_utils_validate_rtc_time()` and add an additional check for the `tm_year` if you wish :) However, if a user tries to set the clock to a value outside of 2000->2099, something has probably gone terribly wrong already :) I would keep it simple and just use the common check.

Reply:
Yeah, I guess I should have been more clear here,
the util function checks the year value to be between 0 and 199.
However, the IRTC register can be +/- 128 or absolute 256.
Which means once passing the struct of year value = 200.
The Util function will return false without me knowing that the year value is
what caused the function for return false, meaning I have to check all values again anyway.

Reply:
That is part of my argument :) passing the year 200 would be year 2100, which is in 65 years :D anyone who will intentionally set the time to a value after 2100 has likely not been born yet :D

Reply:
I added this change to the recent update,
but I claim no responsibility if a value above
200 is set on a critically locked system.